### PR TITLE
Homematic: Added virtual datapoint for HM-Dis-WM55 displays

### PR DESCRIPTION
**Pull request by gerrieg**
Signed-off-by: Gerhard Riegler <gerhard.riegler@gmail.com>

### CODE REVIEW

**Comment by MHerbst** [comment]
If would be good to add support for the E-Paper Display. Most things are working out-of-the-box with this version. See the my detailed comment: https://github.com/openhab/openhab2-addons/issues/1874#issuecomment-285965171

**Comment by gerrieg** [comment]
@MHerbst 
I added led and beeper support for HM-Dis-EP-WM55, maybe you can test if it works? Thanks!

**Comment by MHerbst** [comment]
@gerrieg great! I will probably test it this weekend.

**Comment by MHerbst** [comment]
@gerrieg I was able to perform some tests and I found some problems. 

For my first test, I wrote a rule as follows:
```
rule "Display Test"
when
    Item Dt_1_PressShort received update ON
then
    logInfo("taster", "gedrueckt")
    sendCommand(Dt_3_DisplayLine1, "Line1 a")
    sendCommand(Dt_3_DisplayLine2, "Line2 1")
    sendCommand(Dt_3_DisplayLine3, "Line15")
    sendCommand(Dt_3_DisplaySubmit, ON)
end
```
With this trigger the rule only worked sometimes. Most of the time I got an exception similar to this one:
```
16:02:57.737 [INFO ] [marthome.event.ItemStateChangedEvent] - Dt_1_PressShort changed from ON to OFF
16:02:58.205 [ERROR] [ematic.handler.HomematicThingHandler] - -1 Failure (sending setValue()
NEQ1596559:3
SUBMIT
0x02,0x0a,0x12,0x4c,0x69,0x6e,0x65,0x31,0x20,0x35,0x11,0x80,0x13,0x84,0x0a,0x12,0x20,0x11,0x80,0x0a,0x12,0x20,0x11,0x80,0x0a,0x03
)
org.openhab.binding.homematic.internal.communicator.client.UnknownRpcFailureException: -1 Failure (sending setValue()
NEQ1596559:3
SUBMIT
0x02,0x0a,0x12,0x4c,0x69,0x6e,0x65,0x31,0x20,0x35,0x11,0x80,0x13,0x84,0x0a,0x12,0x20,0x11,0x80,0x0a,0x12,0x20,0x11,0x80,0x0a,0x03
)
        at org.openhab.binding.homematic.internal.communicator.parser.RpcResponseParser.parse(RpcResponseParser.java:46)[150:org.openhab.binding.homematic:2.1.0.201703191319]
        at org.openhab.binding.homematic.internal.communicator.client.BinRpcClient.sendMessage(BinRpcClient.java:90)[150:org.openhab.binding.homematic:2.1.0.201703191319]
        at org.openhab.binding.homematic.internal.communicator.client.BinRpcClient.sendMessage(BinRpcClient.java:78)[150:org.openhab.binding.homematic:2.1.0.201703191319]
        at org.openhab.binding.homematic.internal.communicator.client.RpcClient.setDatapointValue(RpcClient.java:280)[150:org.openhab.binding.homematic:2.1.0.201703191319]
```
When this happens the CCU displayed a message about communication problems, too.

After such an exception the thing state changes to `OFFLINE (COMMUNICATION ERROR)` which is not completely true because a button press will be received and handled (and the state changes back to "ONLINE").

Then I changed the trigger rule to
```
    Item Dt_1_PressShort received update OFF
```
This works definitely better. There seems to be a problem if you try to send a command to the device between the events for "button pressed" and "button released".

The beeper command seems to cause problems. It worked at least one time. But when I try it now to send a beeper command, the device signals an incomplete message. Then I have to restart openHAB because even if I try to send a command to set the text (that normally works), I get this error.

Also setting the LED does not work. I will try to debug both problems.

**Comment by MHerbst** [comment]
@kaikreuzer There seems to be a problem with automatic PR build. After @gerrieg's last commit a new build was triggered and a jar file copied to https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/binding/org.openhab.binding.homematic/2.1.0-SNAPSHOT/ . But this file does not contain the latest changes. 

There is one file where you can see very good that something is wrong: @gerrieg added in the second commit some lines add the end of the file `homematic/extra-descriptions.properties`, but these changes were not in the newly created jar file. 

According to the Jenkins git log both commits were added to the build. The timestamps in the jar file seem a bit odd to me: the compiled java class files have a timestamp of 11:21 but the files in folders homematic and META_INF have a timestamp of 14:36. But according to Jenkins the whole build (https://openhab.ci.cloudbees.com/job/PR-openHAB2-Addons/4220/) only took 59s.

BTW: th

**Comment by kaikreuzer** [comment]
@MHerbst No, there is no problem with the PR builds. The last build from this PR was indeed 4220. Unfortunately, the LATEST build that updated the Homematic snapshot was 4224 - this build was triggered by #2000 (which was clearly a mistake). This is the [risk that I had mentioned here](https://community.openhab.org/t/distributing-bindings-through-the-iot-marketplace/24491):

> Note: The shared Maven repo for PRs obviously has the risk that if somebody else includes changes to a certain binding in his PR that the binary jar is changed accordingly - so it can be tricky to tell, what its origin is. I therefore only recommend using these urls for new bindings that are not yet part of the repo - refactorings of existing bindings are probably better offered elsewhere.

**Comment by gerrieg** [comment]
@MHerbst Thank you for testing, i hope you find the problem, otherwise i will remove the LED an Beeper code.

**Comment by MHerbst** [comment]
@kaikreuzer Thanks for the clarification. I was confused by the git output in the Jenkins protocol.

**Comment by MHerbst** [comment]
@gerrieg The beeper is working now. It just needs some more codes to be sent to the device. I also figured out how the LED is working. To make it flash it is necessary to send the beeper command codes too (with beeper set to "OFF" if you only want to LED to flash).

According to the sample script it seems to be possible to set the number of beeps and a duration. But this did not work in my first test. I will test this again. If it works we would need an additional channel to support the feature.
BTW: the EP device does not support colored text so it would be better to not create channels for line colors.

**Comment by MHerbst** [comment]
@gerrieg Maybe I found a bug in the handling of the button press. I have created a simple rule:
```
rule "Display Test"
when
    Item homematic_HM_Dis_EP_WM55_hm_NEQ1596559_1_PRESS_SHORT received update OFF
then
    logInfo("Display Test", "received off")
end
```
When I start openHAB the script is executed two times without pressing the button. Here is an excerpt from the log:
```
2017-03-21 21:14:21.549 [DEBUG] [omematicDeviceDiscoveryService:89   ] - Homematic device discovery scan finished
2017-03-21 21:14:21.549 [INFO ] [ome.event.ThingStatusInfoEvent:43   ] - 'homematic:bridge:hm' updated: ONLINE
2017-03-21 21:14:21.549 [DEBUG] [.c.thing.internal.ThingManager:598  ] - Register and initialize child 'homematic:HM-Dis-EP-WM55:hm:NEQ1596559' of bridge 'homematic:bridge:hm'.
2017-03-21 21:14:21.549 [INFO ] [nt.ThingStatusInfoChangedEvent:43   ] - 'homematic:bridge:hm' changed from INITIALIZING to ONLINE
2017-03-21 21:14:21.551 [DEBUG] [.c.thing.internal.ThingManager:578  ] - Calling 'HomematicThingHandlerFactory.registerHandler()' for thing 'homematic:HM-Dis-EP-WM55:hm:NEQ1596559'.
2017-03-21 21:14:21.556 [DEBUG] [.c.thing.internal.ThingManager:695  ] - Calling initialize handler for thing 'homematic:HM-Dis-EP-WM55:hm:NEQ1596559' at 'org.openhab.binding.homematic.handler.HomematicThingHandler@3b59a1cf'.
2017-03-21 21:14:21.556 [INFO ] [ome.event.ThingStatusInfoEvent:43   ] - 'homematic:HM-Dis-EP-WM55:hm:NEQ1596559' updated: INITIALIZING
2017-03-21 21:14:21.556 [INFO ] [nt.ThingStatusInfoChangedEvent:43   ] - 'homematic:HM-Dis-EP-WM55:hm:NEQ1596559' changed from UNINITIALIZED (HANDLER_MISSING_ERROR) to INITIALIZING
2017-03-21 21:14:21.557 [DEBUG] [o.o.b.h.i.c.HomematicGateway  :485  ] - Loading values for channel HmChannel[number=0,type=MAINTENANCE,initialized=false] of device 'NEQ1596559'
...
2017-03-21 21:14:21.585 [INFO ] [ome.event.ThingStatusInfoEvent:43   ] - 'homematic:HM-Dis-EP-WM55:hm:NEQ1596559' updated: ONLINE
...
2017-03-21 21:14:21.590 [DEBUG] [.o.b.h.h.HomematicThingHandler:142  ] - Channel linked 'homematic:HM-Dis-EP-WM55:hm:NEQ1596559:1#PRESS_SHORT' from thing id 'NEQ1596559'
2017-03-21 21:14:21.590 [DEBUG] [o.o.b.h.i.c.HomematicGateway  :485  ] - Loading values for channel HmChannel[number=1,type=KEY,initialized=false] of device 'NEQ1596559'
2017-03-21 21:14:21.597 [DEBUG] [.o.b.h.h.HomematicThingHandler:142  ] - Channel linked 'homematic:HM-Dis-EP-WM55:hm:NEQ1596559:1#PRESS_LONG' from thing id 'NEQ1596559'
2017-03-21 21:14:21.597 [DEBUG] [.o.b.h.h.HomematicThingHandler:142  ] - Channel linked 'homematic:HM-Dis-EP-WM55:hm:NEQ1596559:2#PRESS_LONG' from thing id 'NEQ1596559'
2017-03-21 21:14:21.598 [DEBUG] [o.o.b.h.i.c.HomematicGateway  :485  ] - Loading values for channel HmChannel[number=2,type=KEY,initialized=false] of device 'NEQ1596559'
2017-03-21 21:14:21.602 [DEBUG] [.o.b.h.h.HomematicThingHandler:142  ] - Channel linked 'homematic:HM-Dis-EP-WM55:hm:NEQ1596559:2#PRESS_SHORT' from thing id 'NEQ1596559'
...
2017-03-21 21:14:21.610 [INFO ] [nt.ThingStatusInfoChangedEvent:43   ] - 'homematic:HM-Dis-EP-WM55:hm:NEQ1596559' changed from INITIALIZING to ONLINE
...
2017-03-21 21:14:21.613 [INFO ] [smarthome.event.ItemStateEvent:43   ] - homematic_HM_Dis_EP_WM55_hm_NEQ1596559_1_PRESS_SHORT updated to OFF
2017-03-21 21:14:21.613 [DEBUG] [.m.r.r.i.engine.RuleEngineImpl:305  ] - Executing rule 'Display Test'
2017-03-21 21:14:21.613 [INFO ] [smarthome.event.ItemStateEvent:43   ] - homematic_HM_Dis_EP_WM55_hm_NEQ1596559_1_PRESS_SHORT updated to OFF
2017-03-21 21:14:21.614 [INFO ] [smarthome.event.ItemStateEvent:43   ] - homematic_HM_Dis_EP_WM55_hm_NEQ1596559_1_PRESS_LONG updated to OFF
2017-03-21 21:14:21.614 [INFO ] [smarthome.event.ItemStateEvent:43   ] - homematic_HM_Dis_EP_WM55_hm_NEQ1596559_1_PRESS_LONG updated to OFF
2017-03-21 21:14:21.614 [DEBUG] [.m.r.r.i.engine.RuleEngineImpl:305  ] - Executing rule 'Display Test'
2017-03-21 21:14:21.614 [INFO ] [smarthome.event.ItemStateEvent:43   ] - homematic_HM_Dis_EP_WM55_hm_NEQ1596559_2_PRESS_LONG updated to OFF
2017-03-21 21:14:21.615 [INFO ] [smarthome.event.ItemStateEvent:43   ] - homematic_HM_Dis_EP_WM55_hm_NEQ1596559_2_PRESS_LONG updated to OFF
2017-03-21 21:14:21.615 [INFO ] [smarthome.event.ItemStateEvent:43   ] - homematic_HM_Dis_EP_WM55_hm_NEQ1596559_2_PRESS_SHORT updated to OFF
2017-03-21 21:14:21.616 [INFO ] [smarthome.event.ItemStateEvent:43   ] - homematic_HM_Dis_EP_WM55_hm_NEQ1596559_2_PRESS_SHORT updated to OFF
2017-03-21 21:14:21.617 [INFO ] [smarthome.event.ItemStateEvent:43   ] - DisplayTaster_3_DisplayIcon1 updated to NULL
2017-03-21 21:14:21.617 [INFO ] [smarthome.event.ItemStateEvent:43   ] - DisplayTaster_3_DisplayIcon1 updated to NULL
2017-03-21 21:14:21.618 [INFO ] [smarthome.event.ItemStateEvent:43   ] - DisplayTaster_3_DisplayIcon2 updated to NULL
2017-03-21 21:14:21.618 [INFO ] [smarthome.event.ItemStateEvent:43   ] - DisplayTaster_3_DisplayLed updated to NULL
2017-03-21 21:14:21.618 [INFO ] [smarthome.event.ItemStateEvent:43   ] - DisplayTaster_3_DisplayIcon2 updated to NULL
2017-03-21 21:14:21.619 [INFO ] [smarthome.event.ItemStateEvent:43   ] - DisplayTaster_3_DisplaySubmit updated to OFF
2017-03-21 21:14:21.619 [INFO ] [smarthome.event.ItemStateEvent:43   ] - DisplayTaster_3_DisplayLed updated to NULL
2017-03-21 21:14:21.619 [INFO ] [smarthome.event.ItemStateEvent:43   ] - DisplayTaster_3_DisplayBeeper updated to NULL
2017-03-21 21:14:21.619 [INFO ] [o.o.b.h.i.c.HomematicGateway  :499  ] - Can not load values for device: NEQ1596559, channel: 4, paramset: VALUES, maybe there are no values available
2017-03-21 21:14:21.620 [INFO ] [smarthome.event.ItemStateEvent:43   ] - DisplayTaster_3_DisplaySubmit updated to OFF
2017-03-21 21:14:21.620 [DEBUG] [o.o.b.h.i.c.HomematicGateway  :485  ] - Loading values for channel HmChannel[number=5,type=KEY,initialized=false] of device 'NEQ1596559'
2017-03-21 21:14:21.620 [INFO ] [smarthome.event.ItemStateEvent:43   ] - DisplayTaster_3_DisplayLine1 updated to NULL
2017-03-21 21:14:21.620 [INFO ] [smarthome.event.ItemStateEvent:43   ] - DisplayTaster_3_DisplayBeeper updated to NULL
2017-03-21 21:14:21.621 [INFO ] [smarthome.event.ItemStateEvent:43   ] - DisplayTaster_3_DisplayLine3 updated to NULL
2017-03-21 21:14:21.621 [INFO ] [smarthome.event.ItemStateEvent:43   ] - DisplayTaster_3_DisplayLine1 updated to NULL
2017-03-21 21:14:21.621 [INFO ] [smarthome.event.ItemStateEvent:43   ] - DisplayTaster_3_DisplayLine2 updated to NULL
2017-03-21 21:14:21.621 [INFO ] [smarthome.event.ItemStateEvent:43   ] - DisplayTaster_3_DisplayIcon3 updated to NULL
2017-03-21 21:14:21.622 [INFO ] [smarthome.event.ItemStateEvent:43   ] - DisplayTaster_3_DisplayLine3 updated to NULL
2017-03-21 21:14:21.622 [INFO ] [.e.s.model.script.Display Test:48   ] - received off
2017-03-21 21:14:21.622 [INFO ] [smarthome.event.ItemStateEvent:43   ] - DisplayTaster_3_DisplayLine2 updated to NULL
2017-03-21 21:14:21.623 [INFO ] [smarthome.event.ItemStateEvent:43   ] - DisplayTaster_3_DisplayIcon3 updated to NULL
2017-03-21 21:14:21.623 [INFO ] [.e.s.model.script.Display Test:48   ] - received off
```
The interesting part is at timestamp 21:14:21.613. There are two messages that the state of homematic_HM_Dis_EP_WM55_hm_NEQ1596559_1_PRESS_SHORT changed to "OFF". The second execution seems to be a bug.

**Comment by gerrieg** [comment]
@MHerbst 
I've update the EP display handling as suggested by you. I would ask you to test it one more time.

Thank you for the hint with the double initialization at startup, i will test it

**Comment by MHerbst** [comment]
@gerrieg Will test it this weekend. I hope I will find out the missing pieces and we can finalize it.

**Comment by MHerbst** [comment]
@gerrieg Now I had enough time to test it again and add some missing features. I made some smaller corrections and added to new channels to set the number of beeps and the interval between two beeps.

I juast attached a zip file that contains the modified files (only class DisplayTextVirtualDatapoint and the extra-descriptions files).
[EP-Display.zip](https://github.com/openhab/openhab2-addons/files/869159/EP-Display.zip)

**Comment by gerrieg** [comment]
Thank you very much @MHerbst for finding the codes for the EP display. I changed the code of the beep counts and intervals a little bit, they are now calculated.
Maybe you can do a quick review, but i think we got it now.

**Comment by MHerbst** [comment]
> Thank you very much @MHerbst for finding the codes for the EP display. I changed the code of the beep counts and intervals a little bit, they are now calculated.

It's definitely better than my very simple solution :-)

> Maybe  you can do a quick review, but i think we got it now.

Yes, I will do it today or tomorrow.

@gerrieg  What do you think of adding this version to the IoT marketplace (https://community.openhab.org/t/distributing-bindings-through-the-iot-marketplace)?

**Comment by gerrieg** [comment]
As far as i understand, this is only for bindings/PR that have not been merged yet or i want to share a test version. The latest version of the Homematic binding is currently available in the snapshot build, so does it make sense to also add the same version to the marketplace?

**Comment by MHerbst** [comment]
As this version has not been merged yet, I think it would make sense if it takes longer until it is finally merged.

> The latest version of the Homematic binding is currently available in the snapshot build, so does it make sense to also add the same version to the marketplace?

It could make the installation easier because you are able to install only the latest version of this binding and don't have to replace everything. 
But I am also not sure about the best way to distribute binding different versions.

File: addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/virtual/DisplayTextVirtualDatapoint.java [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+            this.code = code;
+        }
+
+        protected String getCode() {
+            return code;
+        }
+
+        /**
+         * Returns the color code.
+         */
```
Comment (MHerbst) [inline addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/virtual/DisplayTextVirtualDatapoint.java]: it returns the icon code

File: addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/virtual/DisplayTextVirtualDatapoint.java [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+        private Beeper(String code) {
+            this.code = code;
+        }
+
+        protected String getCode() {
+            return code;
+        }
+
+        /**
+         * Returns the color code.
```
Comment (MHerbst) [inline addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/virtual/DisplayTextVirtualDatapoint.java]: it returns the beeper code

File: addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/virtual/DisplayTextVirtualDatapoint.java [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+        return DATAPOINT_NAME_DISPLAY_SUBMIT;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void initialize(HmDevice device) {
+        boolean isEpDisplay = isEpDisplay(device);
+        if (device.getType().equals(DEVICE_TYPE_STATUS_DISPLAY) || isEpDisplay) {
```
Comment (MHerbst) [inline addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/virtual/DisplayTextVirtualDatapoint.java]: Suggestion: use a method call for both checks

**Comment by gerrieg** [comment]
@MHerbst Thanks for the review and again a big thank you for testing and finding the codes

@kaikreuzer it's ready to merge now, all the crazy codes to control a display are working

File: addons/binding/org.openhab.binding.homematic/README.md [hunk 4]
```diff
@@ -266,6 +266,52 @@ Duplicate options: TONE3 is ignored, because TONE1 is specified previously.
 smarthome send Display_Options "TEXT, TONE1, BLINK_FAST, TONE3"
 ```
 
+#### DISPLAY_SUBMIT
+Adds multiple virtual datapoints to the HM-Dis-WM55 device to easily send colored text and icons to the display
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.homematic/README.md]: please add an empty line for Jekyll.

File: addons/binding/org.openhab.binding.homematic/README.md [hunk 5]
```diff
@@ -266,6 +266,52 @@ Duplicate options: TONE3 is ignored, because TONE1 is specified previously.
 smarthome send Display_Options "TEXT, TONE1, BLINK_FAST, TONE3"
 ```
 
+#### DISPLAY_SUBMIT
+Adds multiple virtual datapoints to the HM-Dis-WM55 device to easily send colored text and icons to the display 
+
+Example: Display text at line 1,3 and 5 when the bottom button on the display is pressed
+- Items
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.homematic/README.md]: please add an empty line for Jekyll.

File: addons/binding/org.openhab.binding.homematic/README.md [hunk 6]
```diff
@@ -266,6 +266,52 @@ Duplicate options: TONE3 is ignored, because TONE1 is specified previously.
...
+    sendCommand(Display_color_1, "NONE")
+    sendCommand(Display_color_3, "RED")
+    sendCommand(Display_color_5, "BLUE")
+
+    sendCommand(Display_submit, ON)
+end
+```
+
 ### Troubleshooting
 **SHORT & LONG_PRESS events of push buttons do not occur on the event bus**
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.homematic/README.md]: please add an empty line for Jekyll.

**Comment by lopez1de** [comment]
For me (using EP version) the device gets UNREACH after some submits. Buttons are still working then. Does anyone else get this issue?

**Comment by MHerbst** [comment]
The device seems to be a bit sensitive (or in other words: I have the feeling that the firmware has got some bugs). I have the same device and had this problem when sending submits too fast. If one message was not completely processed and you send another one it gets stuck. There seems to also some issues when sending messages to enable the LED. 
I think it's the best to create a new issue, because this one is closed. After I am back from vacation I will probably have some time to work on this issue.

**Pull request merged**